pkg/sarif.go

```diff
@@ -0,0 +1,446 @@
+
+	//nolint
+	// Populate the locations.
+	// Note https://docs.github.com/en/code-security/secure-coding/integrating-with-code-scanning/sarif-support-for-code-scanning#result-object
+	// "Only the first value of this array is used. All other values are ignored."
+	if showDetails {
+		for _, d := range details {
+			if !shouldAddLocation(&d, showDetails, logLevel, minScore, score) {
+				continue
+			}
+
+			loc := location{
+				PhysicalLocation: physicalLocation{
+					ArtifactLocation: artifactLocation{
+						URI:       d.Msg.Path,
+						URIBaseID: "%SRCROOT%",
+					},
+				},
+				Message: &text{Text: d.Msg.Text},
+			}
+
+			// Set the region depending on the file type.
+			loc.PhysicalLocation.Region = detailToRegion(&d)
+			locs = append(locs, loc)
+		}
+	}
+	return locs
+}
+
+func addDefaultLocation(locs []location) []location {
+	// No details or no locations.
+	// For GitHub to display results, we need to provide
+	// a location anyway, regardless of showDetails.
+	detaultLine := 1
+	if len(locs) == 0 {
+		loc := location{
+			PhysicalLocation: physicalLocation{
+				ArtifactLocation: artifactLocation{
+					URI:       ".github/scorecard.yml",
+					URIBaseID: "%SRCROOT%",
+				},
+				Region: region{
+					// TODO: set the line to the check if it's overwritten,
+					// or to the global policy.
+					StartLine: &detaultLine,
+				},
+			},
+		}
+		locs = append(locs, loc)
+	}
+	return locs
+}
+
+func detailsToRelatedLocations(details []checker.CheckDetail,
+	showDetails bool, logLevel zapcore.Level, minScore, score int) []relatedLocation {
+	rlocs := []relatedLocation{}
+
+	//nolint
+	// Populate the related locations.
+	for i, d := range details {
+		if !shouldAddRelatedLocation(&d, showDetails, logLevel, minScore, score) {
+			continue
+		}
+		// TODO: logical locations.
+		rloc := relatedLocation{
+			ID:      i,
+			Message: text{Text: d.Msg.Text},
+			PhysicalLocation: physicalLocation{
+				ArtifactLocation: artifactLocation{
+					// Note: We don't set
+					// URIBaseID: "PROJECTROOT" because
+					// the URL is an `absolute` path.
+					URI: d.Msg.Path,
+				},
+			},
+		}
+		// Set the region depending on file type.
+		rloc.PhysicalLocation.Region = detailToRegion(&d)
+		rlocs = append(rlocs, rloc)
+	}
+	return rlocs
+}
+
+// TODO: update using config file.
+func scoreToLevel(score int) string {
+	// Any of error, note, warning, none.
+	switch score {
+	default:
+		return "error"
+	case checker.MaxResultScore, checker.InconclusiveResultScore:
+		return "note"
+	}
+}
+
+func createSARIFHeader(url, category, name, version string, t time.Time) sarif210 {
+	return sarif210{
+		Schema:  "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
+		Version: "2.1.0",
+		Runs: []run{
+			run{
+				Tool: tool{
+					Driver: driver{
+						Name:           strings.Title(name),
+						InformationURI: url,
+						SemVersion:     version,
+						Rules:          make([]rule, 1),
```

Comment:
Is it intentional that we create a slice with a single empty `rule` here?  This gets overridden later anyway right? So we can just skip this line completely.

Reply:
you're right. Fixed.